policies/recipes/artifacts_test.go

```diff
@@ -0,0 +1,302 @@
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%s/agent.deb", server.URL),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if !exists("/tmp/id1.deb") {
+		t.Errorf("expected fetched artifacts")
+	}
+
+	if err = os.Remove("/tmp/id1.deb"); err != nil {
+		t.Fatalf("could not clean up test data")
+	}
+
+	if art["id1"] != "/tmp/id1.deb" {
+		t.Errorf("expected file entry, but not present")
+	}
+}
+
+func TestFetchArtifacts_http_InvalidURL(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%sagent.deb", server.URL),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpected artifact downloaded")
+	}
+
+	if err == nil || !strings.Contains(err.Error(), "Could not parse url") {
+		t.Errorf("expected url error, found(%+v)", err)
+	}
+
+}
+
+func TestFetchArtifacts_http_NonHTTPRemote(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+
+	url := strings.Replace(server.URL, "http", "ftp", -1)
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%s/agent.deb", url),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
```

Comment:
question: will this unit test succeed if run on a read-only filesystem

Reply:
i have refactored out the code so that in our tests; we do not have to interact with the filesystem.

however, i would still like to know, is that even a valid scenario? i am expecting people to run on their workstation. generally /tmp is writable

Reply:
filesystems are external systems as much as networks and network services are; you should not make any assumptions about their character in tests

Reply:
ok